docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,492 @@
+    - Provision a new machine at the correct version
+    - Poll on health of replicas with the current config hash
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (for an AWS cluster, the image is stored in InfrastructureTemplate.Spec.AMI.ID).
+- Where necessary, we will import behaviours from the MachineSet and MachineDeployment controllers.
+- The controller should tolerate the manual removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+
+##### Control plane  healthcheck
+
+- Will be used during scaling and upgrade operations.
+- There will also be a signal to indicate a healthcheck that could not be made, e.g. due to a network partition.
+- For stacked control planes, we will present etcd quorum status within the Status.Ready field, and also report the number of active cluster members through "ReadyReplicas"
+- The addition of status fields will be done conservatively to prevent them being relied upon even as we choose to deprecate them in the future.
+- Unlike previous lifecycle managers (see Alternatives), ComponentStatus will be not be used as it is deprecated.
+- Examples of specific checks we may perform include:
+  - Checking the managed cluster has a kubeadm-config configmap, which will mark the cluster as Initialized
+  - Checking the managed cluster has a kubelet-$version-configmap matching the desired
+     version to check the state of upgrades
+  - Checking the status of all pods marked with the system-cluster-critical priorityClass within the managed cluster, and providing a count of ready vs total.
+  - Running PodExec etcdctl, or port-forwarding to etcd to get etcd cluster health information
+  - Checking the status of pods marked with the app=kube-dns label as a proxy for information about the health of CNI.
+
+##### Adoption of pre-v1alpha3 Control Plane Machines
+
+- Existing Control Plane Machines will need to be updated with labels matching the expected label selector.
+- The KubeadmConfigSpec can be re-created from the referenced KubeadmConfigs for the Machines matching the label selector.
```

Comment:
This is a little confusing, mind to reword a little?

Reply:
hmm, i think i have too much context... can you be specific about which part is confusing so i can wordsmith?

Reply:
Well, reading it I don't immediately understand the `re-created from the referenced KubeadmConfigs`. Recreated for what purpose? Which `KubeadmConfigSpec` are we referring to and which referenced `KubeadmConfigs` ?

Reply:
awesome, thanks for the help, let me see what i can do!

Reply:
We'll take the v1a2 control plane machines, look up their bootstrap refs, and then try to create controlPlane.spec.kubeadmConfigSpec from them.

Reply:
How does this sound?

> - The adoptee's KubeadmConfigSpec will need to be regenerated based on the adopter's KubeadmConfigs.

Reply:
@vincepri bump^